Project

Sir-Teo opened the pull request:
created a new repository

Sir-Teo committed:
Update README.md

Sir-Teo committed:
Update README.md

Sir-Teo committed:
The skeleton of the project

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
finished a demo

bwiedermann commented on code/circuit.py:
```diff
@@ -0,0 +1,24 @@
+class circuit:
+
+    # a list of lines. The line should be in order
+    connection = []
```
You can delete this line of code because Python is a little different than Java. In Python, we don't have to declare or initialize the `fields`* outside of the constructor. Instead, we just initialize them in the constructor (as you have done), and that's all we need to do.

* Things like `connection` are where an object stores its data. When an object-oriented programmer wants to talk about something like `connection`--but the programmer doesn't want to use terminology from a specific programming language--the programmer will often say "field" or "property" or "member".  When talking about something like `connection` In _Java_, people sometimes use the term "private variables" or  "private data members". That's because, in Java, we often use a private variable to implement a field. But I've avoided using the term "private variable" in my comments above because Python doesn't have private variables.

bwiedermann commented on code/circuit.py:
```diff
@@ -0,0 +1,24 @@
+
+    #Constructors
+    def __init__(self):
+        self.connection = []
```
Keep this line. This line is all you need to declare and initialize a field in Python.

bwiedermann commented on code/circuit.py:
```diff
@@ -0,0 +1,24 @@
+class circuit:
```
Usually, we name a class using upper-case letters, like so: `Circuit`. There's nothing wrong with naming the class `circuit`, though. Just be aware that it is different from what most Python code would do. 

If you choose to rename the class to `Circuit`, you _do not_ need to rename the file. Unlike Java, Python class names do not need to match the file name.

bwiedermann commented on code/element.py:
```diff
@@ -0,0 +1,19 @@
+class element:
+
+    label = "" #label is what the user want to be appear on the circuit, it may contains values e.g. 13nC
+    name = "" #name is the element itself
```
As in `circuit`, you don't need these lines. You can delete them (and move the documentation to the constructor).

bwiedermann commented on code/line.py:
```diff
@@ -0,0 +1,18 @@
+class line:
+
+    elements = []
+
```
As in circuit, you don't need this line. You can delete it.

Sir-Teo committed:
The skeleton of the project is complete

Sir-Teo committed:
unable to implement line class

June 24, 8:51PM

Sir-Teo committed:
first version

Sir-Teo committed:
more changes

Sir-Teo committed:
Delete .DS_Store

Sir-Teo committed:
V2

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
fixed

Sir-Teo committed:
it works

it works. However, I still have to use the __str__ method in element.py, because it is a different object from the circuit one

Sir-Teo committed:
structure

Sir-Teo committed:
parser

Sir-Teo committed:
Error: l1 unreference, trying to come out a solution

Sir-Teo committed:
resolved the unreference error by casting l1 global

Sir-Teo committed:
IT WORKS!

Sir-Teo committed:
updated DSL how to use

Sir-Teo committed:
the program can now support multiple lines!

bwiedermann commented on README.md:
````diff
@@ -1,4 +1,36 @@
+To run this program, first install some packages on your computer. Open the terminal,
+
+```
+$ pip3 install lark-parser
````
Also add: SchemDraw:
`pip3 install lark-parser SchemDraw`

bwiedermann commented on code/grammar.txt:
```diff
@@ -0,0 +1,20 @@
+
+
+    %import common.WS
+    %ignore WS
```
This grammar works great. I'm guessing that you probably adapted it from an existing example in the Lark documentation, which is a good idea.

However, at this stage in your project--when you're about to add more features--I suggest revising the grammar so that it uses more words associated with your domain. "Noun" and "verb" are generic language terms; you want domain-specific ones. Something like:

```
connection: "connect" components "in" CONNECTION_TYPE

// A typical grammar rule for describing a list of things. In this case, we're
// describing a list of components, separated by the word "and"
components: COMPONENT | components "and" COMPONENT

COMPONENT: "capacitor" | "resistor" | "battery"
CONNECTION_TYPE: "series" | "parallel"
``` 

I think that using domain-specific terms in your grammar will make it a lot easier to modify the grammar as you add new features.

Sir-Teo committed:
Update README.md

Sir-Teo committed:
updated grammar(with new labeling features), but it is not working

Sir-Teo committed:
new feature, the user can label elements!

updated grammar file and README file

Sir-Teo committed:
successfully extract information using a transformer

Sir-Teo committed:
process tree

Sir-Teo committed:
it works!

Sir-Teo committed:
updated README.md

Sir-Teo committed:
example computations

harrylyf commented:
Critiques for Notebook 1:
I really like your circuit program! It has a great potential. You can add more features from connect-in-parallel to a completed circuit. And you can also add more elements such as inductors and batteries. One good way to evaluate your language is to ask your physics teacher at Berkley. Your can get the most professional answer from your domain. So for your questions about grammar, I think your current grammar works and esay for me to understand. I did have an idea on improving the grammar, but I am not sure if it is workable. Like we did in Picobot, you can have your define element part be shorter. Like ```set_label capacitor tyler 20000 C ``` Users don't even need to write the unit because once *capacitor* is defined, the farad should be set default. (If you want to represent capacitance, it is better not to put **c** behind the number because I thought it was Coulomb. It's better to put unit behind the number.) Another thing is there is an app called **iCircuit** on Mac. It is quite similar to your language. Hope it helps. Good Work!
Harry

Sir-Teo committed:
improve the way iterating through the parse tree

Sir-Teo committed:
The program can now display greek letters

Sir-Teo committed:
reorganizing the parse tree, so that the data structure is more organized

Sir-Teo committed:
The program works again, but error handling is not functioning well

harrylyf commented:
Critiques for Notebook 2:
Maybe after you have done the parallel connection, you could try to combine parallel and series together in one circuit. And you can have an output window showing the current voltage, current, resistance within each section(components).  However, the expression you have made in *readme.md* confuses me: the resistor alias does not match. Beyond that, the example circuit is in good shape.
Harry

Sir-Teo committed:
some progress made to this new feature

Sir-Teo committed:
small bug fixed

Sir-Teo committed:
finished implementing the "front", my next step is the "back"

harrylyf commented:
Critiques for Notebook 3:
In order to describe the logic gates, I think the best way is call which gates first so that the parser can expect what to parse next. Such as, ```AND A B -> AB``` and ```NOT A -> A*```. I think this will work. I found [this website](http://www.ee.surrey.ac.uk/Projects/CAL/digital-logic/gatesfunc/index.html) pretty informative, maybe containing some interesting sources you might need in the future. For the problem you are solving, I think [this](https://cdelker.bitbucket.io/SchemDraw/SchemDraw.html) may also help. I am not sure how python parser works, but in Scala, if I want to have multiple continuous output, like 
```python
d.add(e.CAP, d='down', botlabel='0.1$\mu$F') 
d.add(e.LINE, to=V1.start) 
d.add(e.GND) 
``` 
the best way if to define a function which combines all these command and make the parser match this function. Hope it help!

Harry

Sir-Teo committed:
found a way to connect in parallel

Sir-Teo committed:
updated circuit.py

Sir-Teo committed:
it works

Sir-Teo committed:
building a new feature, adding elements

Sir-Teo committed:
small bug fixed

Sir-Teo committed:
about to complete this new feature, but still i cant find a way to add multiple elements

Sir-Teo committed:
some progress

Sir-Teo committed:
more simplicity in parser.py/ updating the parsing process so that i can implement the new feature easier

Sir-Teo committed:
Half done of the new feature! a stable version of Add_parallel

Sir-Teo committed:
it works!

Sir-Teo committed:
updated README

Sir-Teo committed:
Update README.md

Sir-Teo committed:
updated README

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
fixed small error

harrylyf commented:
Critiques for Notebook 4:
It is really hard for me to think of another feature for your language because I feel that you basically have covered most of the necessary parts. One of the potential features you can add is calculation: (if the circuit is determined to be a closed circuit) calculate the current and voltage in each component. Another potential work you can do is adding more circuit component to your language: LED, inductor, light, etc.

Harry

Sir-Teo committed:
some prograss

Sir-Teo committed:
added a special method to print out data

harrylyf commented:
Critiques for Notebook 5:
So a circuit that is not closed is an open circuit? That what I thought your question means. So, I think the user needs to define whether the circuit is close or open at the very beginning so that your program will have time to identify if every input components meet the expectation. ```# closed circuit``` 

If the program successfully identifies the circuit to be closed, you can have an output message showing some calculation the program made in response to the given circuit.

Harry

Sir-Teo committed:
new grammar! setting a mode

Sir-Teo committed:
some progress with the new feature

Sir-Teo committed:
finished overriding the element class

Sir-Teo committed:
finished a part of transformer

Sir-Teo committed:
Merge branch 'calculator'

Sir-Teo committed:
Update README.md

Sir-Teo committed:
finished another feature

Sir-Teo committed:
updated examples

Sir-Teo committed:
BUG: cannot add same conponent due to the use of dictionary, other features works well

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
working on a GUI to simplify user experience

Sir-Teo committed:
the gui worked, but unable to render multiple times

Sir-Teo committed:
Update README.md

Sir-Teo committed:
updated images for illustration

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
fixed minor errors

Sir-Teo committed:
improved gui

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
uploaded setup.py and requirement.txt

Sir-Teo committed:
a simple file management system

Sir-Teo committed:
fixed a small bug: unable to update image

harrylyf commented:
Critiques for Notebook 6:
In your GUI version, you can make an input box on the left where users can write codes. Also, you could try if you could use the symbols to represent unit Omega (maybe a library could do this). Overall, I think you have already plenty of useful and powerful features. 

Harry

Sir-Teo committed:
some final changes

Sir-Teo committed:
more elements

Sir-Teo committed:
a bunch of examples

Sir-Teo committed:
Delete .DS_Store

Sir-Teo committed:
Delete .DS_Store

Sir-Teo committed:
Delete .DS_Store

Sir-Teo committed:
updated an video link

Sir-Teo committed:
fixed small bug

Sir-Teo committed:
Update README.md

Sir-Teo committed:
Update README.md

Sir-Teo committed:
smaller gui image

Sir-Teo committed:
Merge branch 'master' of https://github.com/Sir-Teo/project

Sir-Teo committed:
Update README.md

Sir-Teo committed:
Update README.md

Sir-Teo committed:
grammar fixed
